### PR TITLE
SONAR-9616 Use single ES query in api/project_branches/list

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java [hunk 1]
```diff
@@ -360,6 +365,7 @@ public SearchResponse search(IssueQuery query, SearchOptions options) {
 
     configureStickyFacets(query, options, filters, esQuery, requestBuilder);
     requestBuilder.setFetchSource(false);
+    System.out.println(requestBuilder);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS106)

File: server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java [hunk 2]
```diff
@@ -688,22 +693,53 @@ private BoolQueryBuilder createBoolFilter(IssueQuery query) {
...
+            .must(termQuery(IssueIndexDefinition.FIELD_ISSUE_BRANCH_UUID, branchUuid))
+            .must(termQuery(IssueIndexDefinition.FIELD_ISSUE_IS_MAIN_BRANCH, Boolean.toString(false))))
+          .subAggregation(
+            AggregationBuilders.terms("types").field(IssueIndexDefinition.FIELD_ISSUE_TYPE)));
+    });
+    SearchResponse response = request.get();
+    return response.getAggregations().asList().stream()
+      .map(x -> (InternalFilter) x)
+      .flatMap(bucket -> {
+        Map<String, Long> typesCount = ((StringTerms) bucket.getAggregations().get("types")).getBuckets().stream()
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java]: MINOR Remove this unnecessary cast to "StringTerms". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1905)

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java [hunk 3]
```diff
@@ -153,19 +152,14 @@ private static void setLongLivingBranchStatus(WsBranches.Branch.Builder builder,
     }
   }
 
-  private void setShortLivingBranchStatus(WsBranches.Branch.Builder builder, BranchDto branch) {
+  private void setShortLivingBranchStatus(WsBranches.Branch.Builder builder, BranchDto branch, BranchStatistics branchStatistics) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java]: MAJOR Make "setShortLivingBranchStatus" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L130](https://github.com/SonarSource/sonarqube/blob/5e9606e36518842e4171f480a54655f5bbc30336/server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java#L130): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java [hunk 4]
```diff
@@ -112,19 +106,24 @@ public void handle(Request request, Response response) throws Exception {
...
       branches.stream()
         .filter(b -> b.getKeeType().equals(BranchKeyType.BRANCH))
-        .forEach(b -> addBranch(protobufResponse, b, mergeBranchesByUuid, qualityGateMeasuresByComponentUuids.get(b.getUuid())));
+        .forEach(b -> addBranch(protobufResponse, b, mergeBranchesByUuid, qualityGateMeasuresByComponentUuids.get(b.getUuid()), branchStatisticsByBranchUuid.get(b.getUuid())));
       WsUtils.writeProtobuf(protobufResponse.build(), request, response);
     }
   }
 
-  private void addBranch(WsBranches.ListWsResponse.Builder response, BranchDto branch, Map<String, BranchDto> mergeBranchesByUuid, @Nullable MeasureDto qualityGateMeasure) {
+  private void addBranch(WsBranches.ListWsResponse.Builder response, BranchDto branch, Map<String, BranchDto> mergeBranchesByUuid, @Nullable MeasureDto qualityGateMeasure,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java]: MAJOR Make "addBranch" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L130](https://github.com/SonarSource/sonarqube/blob/9dcecd7641b1c3f4596aa76e1277d32aad6fc699/server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java#L130): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L130](https://github.com/SonarSource/sonarqube/blob/33bc4848a0c305e2037bbb7f57760ae209d00f4d/server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java#L130): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java [hunk 5]
```diff
@@ -688,22 +693,53 @@ private BoolQueryBuilder createBoolFilter(IssueQuery query) {
...
+      return Collections.emptyList();
+    }
+
+    SearchRequestBuilder request = client.prepareSearch(IssueIndexDefinition.INDEX_TYPE_ISSUE)
+      .setQuery(
+        boolQuery()
+          .must(termsQuery(IssueIndexDefinition.FIELD_ISSUE_BRANCH_UUID, branchUuids))
+          .must(termQuery(IssueIndexDefinition.FIELD_ISSUE_IS_MAIN_BRANCH, Boolean.toString(false))))
+      .setSize(0);
+    IntStream.range(0, branchUuids.size()).forEach(i -> {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java]: Could be improved by using a single aggregation:
```
SearchRequestBuilder request = client.prepareSearch(IssueIndexDefinition.INDEX_TYPE_ISSUE)
      .setQuery(
        boolQuery()
          .must(termsQuery(IssueIndexDefinition.FIELD_ISSUE_BRANCH_UUID, branchUuids))
          .must(termQuery(IssueIndexDefinition.FIELD_ISSUE_IS_MAIN_BRANCH, Boolean.toString(false))))
      .setSize(0);
    request
      .addAggregation(AggregationBuilders.terms("branchUuid").field(IssueIndexDefinition.FIELD_ISSUE_BRANCH_UUID)
        .subAggregation(
          AggregationBuilders.terms("types").field(IssueIndexDefinition.FIELD_ISSUE_TYPE)));
```

File: server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java [hunk 6]
```diff
@@ -688,22 +690,46 @@ private BoolQueryBuilder createBoolFilter(IssueQuery query) {
...
+          .must(termsQuery(IssueIndexDefinition.FIELD_ISSUE_BRANCH_UUID, branchUuids))
+          .must(termQuery(IssueIndexDefinition.FIELD_ISSUE_IS_MAIN_BRANCH, Boolean.toString(false))))
+      .setSize(0)
+      .addAggregation(AggregationBuilders.terms("branchUuids")
+        .field(IssueIndexDefinition.FIELD_ISSUE_BRANCH_UUID)
+        .size(branchUuids.size())
+        .subAggregation(AggregationBuilders.terms("types")
+          .field(IssueIndexDefinition.FIELD_ISSUE_TYPE)));
+    SearchResponse response = request.get();
+    return ((StringTerms) response.getAggregations().get("branchUuids")).getBuckets().stream()
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java]: MINOR Remove this unnecessary cast to "StringTerms". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1905)

File: server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java [hunk 7]
```diff
@@ -688,22 +690,46 @@ private BoolQueryBuilder createBoolFilter(IssueQuery query) {
...
+      .map(bucket -> new BranchStatistics(bucket.getKeyAsString(),
+        ((StringTerms) bucket.getAggregations().get("types")).getBuckets()
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndex.java]: MINOR Remove this unnecessary cast to "StringTerms". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1905)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L130](https://github.com/SonarSource/sonarqube/blob/af4f19f917983f4d50d40dfde16af4b27121897e/server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/ListAction.java#L130): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**